### ISSUE TITLE
Add a user-funcallable-class?

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#2537]
**Is your feature request related to a problem? Please describe.**

We have configurable `user-class`-es for almost everything in Nyxt. `command` and `internal-page` are classes too, although not `user-class`-es. We may bump into this at some point, requiring to configure some part of the internal page, for example.

It would be nice if this configuration would work by the same rules as the configuration of regular user classes, with `define-configuration` and `customize-instance` as the convenient ways to hook into commands and change their behavior.

**Describe the solution you'd like**

Introduce `user-funcallable-class` and make e.g. `command` and `internal-page` to inherit from it.

**Describe alternatives you've considered**

Well, both `command` and `internal-page` inherit from generics and are non-specialized method by default, which makes them hookable with `:before`, `:after`, and `:around` methods and `intitialize-instance` methods too. It's a lot of configuration space. For example, all the settings can be tweaked in `initialize-instance :after` of a command (by the way, should we use `cascade` method combination for those too, so that parent methods are invoked too?). But still, having the proper and `define-configuration`-friendly way to configure those would be much better and cleaner than overriding some `initialize-instance` or something.

**Additional context**

I've tried to implement it myself, adding `user-funcallable-class` (inheriting from `user-class` and `closer-mop:funcallable-standard-class`), and then trying to use it as command metaclass. This caused (reasonable) conflict between the `standard-funcallable-class standard-generic-function` and `user-funcallable-class command`. Adding yet another layer of indirection (`user-generic-function`) didn't help -- `standard-generic-function` is still there in the hierarchy, breaking this fragile construction into pieces. Do we really need `standard-generic-function` there? (҂⌣̀_⌣́)

**Comment by aadcg** [comment]
> It would be nice if this configuration would work by the same rules as the configuration of regular user classes, with define-configuration and customize-instance as the convenient ways to hook into commands and change their behavior.

Seems absolutely reasonable to me. If there's no good reason to break this symmetry, then let's try to preserve it.

Can't give any valuable insight. I'm wondering: what tells apart `user-class`es and `user-funcallable-class`es, in your construction?

**Comment by aartaka** [comment]
> I'm wondering: what tells apart `user-class`es and `user-funcallable-class`es, in your construction?

The fact that the latter are `funcall`-able :P 

I mean, the general separation of things in MOP seems to be between the `stadard-class`, which all the classes (and objects) descend from, and `standard-funcallable-class`, which all the funcallable classes (=functions) descend from.

**Comment by Ambrevar** [comment]
Seems reasonable to me too!

**Comment by aadcg** [comment]
All clear. [Reading the MOP specification helped](https://clos-mop.hexstreamsoft.com/concepts/#funcallable-instances).

**Comment by aartaka** [comment]
@aadcg, sorry for being opaque there (´･_･`)

**Comment by aartaka** [comment]
> Seems reasonable to me too!

Although it seems to be quite unimplementable. The best one I've got is:
``` lisp
(defclass user-funcallable-class (user-class closer-mop:funcallable-standard-class)
  ()
  (:documentation "The user-class version of `closer-mop:funcallable-standard-class', so that
funcallable objects can be configured too."))
```
used as a meta-class of `command`. And it errors out with opaque and unavoidable:
```
git/nyxt/source/command.lisp:1:1:
  error: 
    The class
    #<SB-MOP:FUNCALLABLE-STANDARD-CLASS SB-MOP:FUNCALLABLE-STANDARD-OBJECT> was
    specified as a superclass of the class
    #<USER-FUNCALLABLE-CLASS NYXT:COMMAND {1008AD5E53}>, but the metaclasses
    #<STANDARD-CLASS SB-MOP:FUNCALLABLE-STANDARD-CLASS> and
    #<STANDARD-CLASS NYXT::USER-FUNCALLABLE-CLASS> are incompatible.
```

**Comment by aadcg** [comment]
Actually, naively, I'd say that `user-funcallable-class` should not inherit from `user-class`. We should rather mimic what we did with `user-class`:

```lisp
(defclass user-funcallable-class (closer-mop:funcallable-standard-class) ...)
```

Which will imply duplication of code perhaps, but is it too bad?

**Comment by aartaka** [comment]
Hmmmmmmmmm. That could work Ｏ(≧▽≦)Ｏ

**Comment by Ambrevar** [comment]
@aartaka What you are seeing is expected indeed, you need to restart your image.  As far as I know, MOP does not allow for changing the metaclasses of live objects.

**Comment by aartaka** [comment]
> @aartaka What you are seeing is expected indeed, you need to restart your image. As far as I know, MOP does not allow for changing the metaclasses of live objects.

No, that wasn't the problem. The problem was that I haven't validated all the metaclasses and used conflicting parents (`standard-class` and `closer-mop:funcallable-standard-class`) in the same class.

**Comment by Ambrevar** [comment]
Works for me with

```lisp
(defmethod closer-mop:validate-superclass ((class user-funcallable-class)
                                           (superclass user-class))
  t)

(defmethod closer-mop:validate-superclass ((class user-funcallable-class)
                                           (superclass closer-mop:funcallable-standard-class))
  t)
```
